python/lsst/cp/pipe/cpFlatNormTask.py

```diff
@@ -74,93 +165,207 @@ class CpFlatNormalizationTask(pipeBase.PipelineTask,
+            Expected model bgMatrix values, calculated from E and G.
+
+        Notes
+        -----
+
+        The set of background measurements B[exposure, detector] of
+        flat frame data should be defined by a "Cartesian" product of
+        two vectors, E[exposure] and G[detector].  The E vector
+        represents the total flux incident on the focal plane.  In a
+        perfect camera, this is simply the sum along the columns of B
+        (np.sum(B, axis=0)).
+
+        However, this simple model ignores differences in detector
+        gains, the vignetting of the detectors, and the illumination
+        pattern of the source lamp.  The G vector describes these
+        detector dependent differences, which should be identical over
+        different exposures.  For a perfect lamp of unit total
+        intensity, this is simply the sum along the rows of B
+        (np.sum(B, axis=1)).
+
+        The algorithm here, from pipe_drivers/constructCalibs.py and
+        from there from Eugene Magnier/PanSTARRS [1]_, attempts to
+        iteratively solve this decomposition from initial "perfect" E
+        and G vectors.  The operation is performed in log space to
+        reduce the multiply and divides to linear additions and
+        subtractions.
+
+        References
+        ----------
+        .. [1] https://svn.pan-starrs.ifa.hawaii.edu/trac/ipp/browser/trunk/psModules/src/detrend/pmFlatNormalize.c  # noqa: E501
+
+        """
+        numExps = bgMatrix.shape[1]
+        numChips = bgMatrix.shape[0]
+
+        logMeas = np.log(bgMatrix)
+        logMeas = np.ma.masked_array(logMeas, ~np.isfinite(logMeas))
+        logG = np.zeros(numChips)
+        logE = np.array([np.average(logMeas[:, iexp] - logG,
+                                    weights=bgCounts[:, iexp]) for iexp in range(numExps)])
+
+        for iter in range(iterations):
+            logG = np.array([np.average(logMeas[ichip, :] - logE,
+                                        weights=bgCounts[ichip, :]) for ichip in range(numChips)])
+
+            bad = np.isnan(logG)
+            if np.any(bad):
+                logG[bad] = logG[~bad].mean()
+
+            logE = np.array([np.average(logMeas[:, iexp] - logG,
+                                        weights=bgCounts[:, iexp]) for iexp in range(numExps)])
+            fluxLevel = np.average(np.exp(logG), weights=np.sum(bgCounts, axis=1))
+
+            logG -= np.log(fluxLevel)
+            self.log.info(f"ITER {iter}: Flux: {fluxLevel}")
+            self.log.info(f"Exps: {np.exp(logE)}")
+            self.log.info(f"{np.mean(logG)}")
+
+        logE = np.array([np.average(logMeas[:, iexp] - logG,
+                                    weights=bgCounts[:, iexp]) for iexp in range(numExps)])
+        bgModel = np.zeros((numChips, numExps))
+        for ichip in range(numChips):
+            for iexp in range(numExps):
+                bgModel[ichip][iexp] = np.exp(logE[iexp] - logG[ichip])
```

Comment:
Iterations are inefficient in python. Can't you write something like `bgModel = np.exp(logE[:, np.newaxis] - logG[np.newaxis, :])` (untested!).

Reply:
I had to switch the axes, but I think that does do the right thing.